app/src/main/kotlin/com/flxrs/dankchat/DankChatViewModel.kt

```diff
@@ -1,31 +1,61 @@
 package com.flxrs.dankchat
 
 import androidx.lifecycle.ViewModel
+import androidx.lifecycle.viewModelScope
 import com.flxrs.dankchat.data.ChatRepository
 import com.flxrs.dankchat.data.DataRepository
+import com.flxrs.dankchat.data.api.ApiManager
+import com.flxrs.dankchat.preferences.DankChatPreferenceStore
+import com.flxrs.dankchat.utils.extensions.withoutOAuthSuffix
 import dagger.hilt.android.lifecycle.HiltViewModel
+import kotlinx.coroutines.channels.BufferOverflow
+import kotlinx.coroutines.flow.MutableSharedFlow
+import kotlinx.coroutines.flow.asSharedFlow
+import kotlinx.coroutines.launch
 import javax.inject.Inject
 
 @HiltViewModel
 class DankChatViewModel @Inject constructor(
     private val chatRepository: ChatRepository,
+    private val dankChatPreferenceStore: DankChatPreferenceStore,
+    private val apiManager: ApiManager,
     dataRepository: DataRepository,
 ) : ViewModel() {
 
     val commands = dataRepository.commands
     var started = false
         private set
 
+    private val _oauthResult = MutableSharedFlow<OAuthResult>(extraBufferCapacity = 2, onBufferOverflow = BufferOverflow.DROP_OLDEST)
+    val oAuthResult get() = _oauthResult.asSharedFlow()
+
     fun init(name: String, oAuth: String, channels: List<String>, tryReconnect: Boolean) {
         if (tryReconnect && started) {
             chatRepository.reconnectIfNecessary()
         } else {
             started = true
-            chatRepository.connectAndJoin(name, oAuth, channels)
+            viewModelScope.launch {
+                val token = oAuth.withoutOAuthSuffix
+                val nameToUse = runCatching {
+                    val result = apiManager.validateUser(token)
+                    if (result == null && token.isNotEmpty()) {
+                        _oauthResult.emit(OAuthResult.OAuthTokenInvalid)
```

Comment:
If the token is invalid we still emit the `OauthResult.UserName` here. I think it would be better to early-return with
```kotlin
return@runCatching null
```